pkg/service/restore_data.go

```diff
@@ -222,36 +229,39 @@ func (r *dataRestorer) restoreByTimeSync(
 	ctx context.Context,
 	request *model.RestoreTimestampRequest,
 	jobID model.RestoreJobID,
-	fullBackupsByNamespace map[string][]model.BackupDetails,
 	logger *slog.Logger,
-) {
+) error {
+	// Lock the routine storage from retention manager for the duration of restore.
+	routineStorageLock := r.routineStorage.Get(request.RoutineName)
+	routineStorageLock.RLock()
+	defer routineStorageLock.RUnlock()
+
+	backupsByNamespace, err := r.findBackupsToRestore(ctx, request)
+	if err != nil {
+		return err
+	}
+
 	client, err := r.clientManager.GetClient(request.DestinationCluster)
 	if err != nil {
-		slog.Error("Failed to restore by timestamp",
-			slog.Any("cluster", request.DestinationCluster.ClusterLabel),
-			attr.Error(err))
-		r.restoreJobs.finishJob(jobID, err)
-		return
+		return fmt.Errorf("failed to get client for cluster %s: %w",
+			util.ValueOrZero(request.DestinationCluster.ClusterLabel), err)
 	}
 	defer r.clientManager.Close(client)
 
-	var wg sync.WaitGroup
-	var multiError error
-	for namespace, nsBackup := range fullBackupsByNamespace {
-		wg.Add(1)
-		go func(namespace string, nsBackup []model.BackupDetails) {
-			defer wg.Done()
+	// Run namespace restores concurrently and collect errors safely.
+	var eg errgroup.Group
+	for namespace, nsBackup := range backupsByNamespace {
+		eg.Go(func() error {
 			if err := r.restoreNamespace(ctx, client, request, jobID, namespace, nsBackup, logger); err != nil {
-				multiError = errors.Join(multiError,
-					fmt.Errorf("failed to restore routine %s, namespace %s by timestamp: %w",
-						request.RoutineName, namespace, err))
+				return fmt.Errorf("failed to restore routine %s, namespace %s by timestamp: %w",
+					request.RoutineName, namespace, err)
 			}
-		}(namespace, nsBackup)
+			return nil
+		})
 	}
 
-	wg.Wait()
-
-	r.restoreJobs.finishJob(jobID, multiError)
+	// errgroup.Wait returns nil if all succeed, otherwise a joined error.
+	return eg.Wait()
```

Comment:
From the `Wait` documentation:
> // Wait blocks until all function calls from the Go method have returned, then
// returns the first non-nil error (if any) from them.